virtualizarr/manifests/store.py

```diff
@@ -137,34 +133,92 @@ def parse_manifest_index(key: str, chunk_key_encoding: str = ".") -> tuple[int,
+            client_options={"allow_http": True},
+            skip_signature=True,
+            virtual_hosted_style_request=False,
+            region=_find_bucket_region(bucket),
+        )
+
+    raise NotImplementedError(f"{parsed.scheme} is not yet supported")
+
+
+def _sort_stores_by_prefix_length(input_dict):
+    sorted_items = sorted(input_dict.items(), key=lambda x: len(x[0]), reverse=True)
+    return dict(sorted_items)
+
+
+class ObjectStoreRegistry:
+    """
+    ObjectStoreRegistry maps URLs to ObjectStore instances, and allows ManifestStores to read from different ObjectStore instances.
     """
-    # Sort keys by length in descending order to ensure longer, more specific matches take precedence
-    sorted_keys = sorted(stores.keys(), key=len, reverse=True)
-
-    # Check each key to see if it's a prefix of the uri_string
-    for key in sorted_keys:
-        if request_key.startswith(key):
-            parsed_key = urlparse(request_key)
-            return StoreRequest(store=stores[key], key=parsed_key.path)
-    # if no match is found, raise an error
-    raise ValueError(
-        f"Expected the one of stores.keys() to match the data prefix, got {stores.keys()} and {request_key}"
-    )
+
+    _stores: dict[str, ObjectStore]
+
+    @classmethod
+    def __init__(self, stores: dict | None = None):
+        stores = stores or {}
+        for store in stores.values():
+            if not store.__class__.__module__.startswith("obstore"):
+                raise TypeError(f"expected ObjectStore class, got {store!r}")
+        self._stores = _sort_stores_by_prefix_length(stores)
+
+    def register_store(self, url: str, store: ObjectStore):
+        """
+        Register a store using the given url
+
+        If a store with the same key existed before, it is replaced
+        """
+        self._stores[url] = store
+        self._stores = _sort_stores_by_prefix_length(self._stores)
+
+    def get_store(self, url: str) -> ObjectStore:
+        """
+        Get a suitable store for the provided URL. For example:
+
+            - URL with scheme file:/// or no scheme will return the default LocalFS store
+            - URL with scheme s3://bucket/ will return the S3 store
+
+        If no `ObjectStore` is found for the `url`, ad-hoc discovery may be executed depending on the
+        `url`. An `ObjectStore` may be lazily created and registered.
+
+        Parameters:
+        -----------
+        url : str
+            A url to identify the appropriate object_store instance
+
+        Returns:
+        --------
+        StoreRequest
+        """
+        # Check each key to see if it's a prefix of the uri_string
+        for prefix, store in self._stores.items():
+            if url.startswith(prefix):
+                # Return an existing configured store and parsed request path
+                return store
+        # Use anonymous default store if not in pre-configured stores
+        prefix, store = _default_object_store(url)
```

Comment:
Another option, if you expect the user to have boto3 in their environment, is to use the default boto3 credentials (via the boto3 credential provider), which will let them access buckets through their default profile automatically.

But it's not great to have different functionality depending on whether boto3 is installed.

Reply:
thanks, we can keep this in mind if needed later on